include/votca/xtp/bseoperator_btda.h

```diff
@@ -0,0 +1,232 @@
+#define __VOTCA_BSEOP_BTDA_H
+#include <votca/xtp/bse_operator.h>
+#include <votca/xtp/eigen.h>
+namespace votca {
+namespace xtp {
+
+template<typename MatrixReplacementA, typename MatrixReplacementB>
+class HamiltonianOperator;
+
+}
+}  // namespace votca
+namespace Eigen {
+namespace internal {
+// MatrixReplacement looks-like a SparseMatrix, so let's inherits its traits:
+template <typename MatrixReplacementA, typename MatrixReplacementB>
+struct traits<votca::xtp::HamiltonianOperator<MatrixReplacementA,MatrixReplacementB>>
+    : public Eigen::internal::traits<Eigen::MatrixXd> {};
+}  // namespace internal
+}  // namespace Eigen
+
+namespace votca {
+namespace xtp {
+template <typename MatrixReplacementA, typename MatrixReplacementB>
+class HamiltonianOperator
+    : public Eigen::EigenBase<HamiltonianOperator<MatrixReplacementA,MatrixReplacementB>> {
+ public:
+  // Required typedefs, constants, and method:
+  typedef double Scalar;
+  typedef double RealScalar;
+  typedef int StorageIndex;
+  enum {
+    ColsAtCompileTime = Eigen::Dynamic,
+    MaxColsAtCompileTime = Eigen::Dynamic,
+    IsRowMajor = false
+  };
+
+  HamiltonianOperator(const MatrixReplacementA &A,
+                      const MatrixReplacementB &B) 
+  : _A(A), _B(B) 
+  { 
+    _size = 2*A.cols() ;
+    _diag = get_diagonal();
+   };
+      
+    
+  class InnerIterator {
+   public:
+    InnerIterator(const HamiltonianOperator& xpr, const Eigen::Index& id)
+        : _xpr(xpr), _id(id){};
+
+    InnerIterator& operator++() {
+      _row++;
+      return *this;
+    }
+    operator bool() const {
+      return _row < _xpr._size;
+    }  
+    
+    double value() const { return _xpr(_row, _id); }
+    Eigen::Index row() const { return _row; }
+    Eigen::Index col() const { return _id; }
+    Eigen::Index index() const { return row(); }
+
+   private:
+    const HamiltonianOperator& _xpr;
+    const Eigen::Index _id;
+    Eigen::Index _row = 0;
+  };
+
+  Eigen::Index rows() const { return this->_size; }
+  Eigen::Index cols() const { return this->_size; }
+  Eigen::Index outerSize() const { return this->_size; }
+
+  template <typename Vtype>
+  Eigen::Product<HamiltonianOperator, Vtype, Eigen::AliasFreeProduct>
+      operator*(const Eigen::MatrixBase<Vtype>& x) const {
+    return Eigen::Product<HamiltonianOperator, Vtype,
+                          Eigen::AliasFreeProduct>(*this, x.derived());
+  }
+
+  // this is not a fast method
+  const double& operator()(const size_t i, const size_t j) const {
+    if (i==j) {
+      return _diag(i);
+    } else {
+      Eigen::RowVectorXd row_out = row(i);
+      return  row_out(j);
+    }
+  };
+
+  //  get a row of the operator
+  Eigen::RowVectorXd row(int index) const {
+    /* Returna row of the operator 
+    H = [ A   B
+         -B* -A* ]
+    */
+    int lsize = this->_size;
+    int halfsize = lsize/2;
+    Eigen::RowVectorXd row_out = Eigen::RowVectorXd::Zero(lsize);
+
+    if (index < halfsize)
+    {
```

Comment:
I do not think we should use the whole row vector formalism here?

Reply:
done as well

Reply:
_A and _B must be public though ...